src/test/java/dev/aherscu/instancio/experiment/PersonRequestHandlerTest.java

```diff
@@ -1,58 +1,77 @@
 
-import dev.aherscu.instancio.experiment.logic.*;
-import dev.aherscu.instancio.experiment.model.*;
-import org.testng.annotations.*;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.both;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasProperty;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.instancio.Select.all;
+import static org.instancio.Select.field;
 
 public class PersonRequestHandlerTest {
 
+    private static Model<CreatePersonRequest> createValidRequestModel() {
+        return Instancio.of(CreatePersonRequest.class)
+                .withSettings(Settings.create()
+                        .set(Keys.BEAN_VALIDATION_ENABLED, true))
+                .toModel();
+    }
+
     @DataProvider
     private static CreatePersonRequest[] data() {
-        return new CreatePersonRequest[] {
-            CreatePersonRequest.builder()
-                .person(Person.builder()
-                    .address(Address.builder()
-                        .line1("1")
-                        .city("A")
-                        .country("A")
-                        .zipCode("1")
-                        .build())
-                    .phone(Phone.builder()
-                        .countryCode(1)
-                        .number(1)
-                        .build())
-                    .build())
-                .build(),
-        };
+        return Instancio.stream(createValidRequestModel())
+                .limit(100)
+                .toArray(CreatePersonRequest[]::new);
     }
 
     @DataProvider
     private static CreatePersonRequest[] negativeData() {
-        return new CreatePersonRequest[] {
-            CreatePersonRequest.builder()
-                .build(),
-            CreatePersonRequest.builder()
-                .person(Person.builder()
-                    .build())
-                .build(),
-            CreatePersonRequest.builder()
-                .person(Person.builder()
-                    .address(Address.builder()
-                        .line1("1")
-                        .city("A")
-                        .country("A")
-                        .zipCode("1")
-                        .build())
-                    .phone(Phone.builder()
-                        .countryCode(10000)
-                        .number(1)
-                        .build())
-                    .build())
-                .build(),
-        };
+        // either empty or a string of 21+ characters in length
+        final StringSpec emptyStringOrLengthOver21 = Gen.string().allowEmpty().minLength(21);
+        final Generator<?> nullGenerator = random -> null;
+
+        // create an otherwise valid object, but with a single invalid field
+        final List<Tuple<TargetSelector, GeneratorSpec<?>>> negativeInputs = Arrays.asList(
```

Comment:
This means that if an annotation has changed then it has to be changed here too.

Reply:
That's true. If an annotation changes, then the invalid input to test that constraint would also change (especially if you're interested in boundary values, that is, the invalid value just outside the valid range).

The linked discussion is for generating valid data only. There's no out-of-the-box support for saying "given these constraints, generate an invalid object". That's somewhat outside the scope of the library, but could probably be implemented as an extension.

Reply:
> The linked discussion is for generating valid data only. There's no out-of-the-box support for saying "given these constraints, generate an invalid object". That's somewhat outside the scope of the library, but could probably be implemented as an extension.

Ok... so how should I approach such "extension".
According to your comment
https://github.com/orgs/instancio/discussions/785#discussioncomment-7409535
The "generator provider SPI" might not be the best option. Then, what else?

Reply:
I'm not really sure off the top of my head. It'll require some experimentation to see what works. Are you open to forking the project to implement this feature?